test/quadiron_c_utest.cpp

```diff
@@ -0,0 +1,246 @@
+        std::vector<int> wanted_idxs(n_outputs);
+
+        for (int i = 0; i < n_data; i++) {
+            data.at(i).resize(block_size + metadata_size);
+            _data[i] = data.at(i).data();
+            ref_data.at(i).resize(block_size);
+            _ref_data[i] = ref_data.at(i).data();
+            randomize_buffer(_data[i] + metadata_size, block_size);
+            std::memcpy(_ref_data[i], _data[i] + metadata_size, block_size);
+        }
+
+        for (int i = 0; i < n_outputs; i++) {
+            parity.at(i).resize(block_size + metadata_size);
+            _parity[i] = parity.at(i).data();
+        }
+
+        // we want all parities
+        for (int i = 0; i < n_outputs; i++) {
+            wanted_idxs[i] = 1;
+        }
+
+        ASSERT_EQ(
+            quadiron_fnt32_encode(
+                inst,
+                _data.data(),
+                _parity.data(),
+                wanted_idxs.data(),
+                block_size),
+            0);
+
+        for (int i = 0; i < n_data; i++) {
+            if (missing_idxs[i]) {
+                std::memset(_data[i], 0, block_size + metadata_size);
+            }
+        }
+
+        for (int i = 0; i < n_parities; i++) {
+            if (missing_idxs[n_data + i]) {
+                std::memset(_parity[i], 0, block_size + metadata_size);
+            }
+        }
+
+        ASSERT_EQ(
+            quadiron_fnt32_decode(
+                inst,
+                _data.data(),
+                _parity.data(),
+                missing_idxs.data(),
+                block_size),
+            0);
+
+        for (int i = 0; i < n_data; i++) {
+            ASSERT_EQ(
+                std::memcmp(_data[i] + metadata_size, _ref_data[i], block_size),
+                0);
+        }
+
+        for (int i = 0; i < n_data; i++) {
+            if (missing_idxs[i]) {
+                ASSERT_EQ(
+                    quadiron_fnt32_reconstruct(
+                        inst,
+                        _data.data(),
+                        _parity.data(),
+                        missing_idxs.data(),
+                        i,
+                        block_size),
+                    0);
+            }
+        }
+
+        for (int i = 0; i < n_parities; i++) {
+            if (missing_idxs[n_data + i]) {
+                ASSERT_EQ(
+                    quadiron_fnt32_reconstruct(
+                        inst,
+                        _data.data(),
+                        _parity.data(),
+                        missing_idxs.data(),
+                        n_data + i,
+                        block_size),
+                    0);
+            }
+        }
+
+        quadiron_fnt32_delete(inst);
+    }
+
+    void test_all_decodable_scenarios(int k, int m, int systematic)
+    {
+        for (int i = 0; i <= m; i++) {
+            auto combinations = generate_combinations(k + m, i);
+            for (auto it = combinations.begin(); it != combinations.end();
```

Comment:
Using auto is clearly better, but it's not the best we can do: use range-based for loop.

```c++
            for (const auto& combination : combinations) {
                std::vector<int> missing_idxs(k + m);
                convert_idx_list(combination, i, missing_idxs, k + m);
                test_encode_decode_reconstruct(
                    k, m, 10000, systematic, missing_idxs);
            }
```